src/refdb_reftable.c

```diff
@@ -0,0 +1,514 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "git2/sys/refdb_backend.h"
+#include "git2/sys/refs.h"
+#include "reflog.h"
+#include "signature.h"
+#include "wildmatch.h"
+#include "../reftable/reftable.h"
+
+typedef struct {
+	git_refdb_backend parent;
+	git_repository *repo;
+	struct stack *stack;
+} reftable_backend;
+
+typedef struct {
+	git_reference_iterator parent;
+	struct merged_table *table;
+	struct iterator iter;
+	const char *glob;
+} reftable_iterator;
+
+typedef struct {
+	struct ref_record *refs;
+	struct log_record *logs;
+	size_t nrefs;
+	size_t nlogs;
+	uint64_t index;
+} reftable_writer_arg;
+
+static int reftable_reference_from_record(git_reference **out, struct ref_record *record)
+{
+	git_reference *ref;
+	git_oid oid, peeled;
+	int error;
+
+	if (record->target) {
+		ref = git_reference__alloc_symbolic(record->ref_name, record->target);
+	} else {
+		if ((error = git_oid_fromraw(&oid, record->value)) < 0 ||
+		    (error = git_oid_fromraw(&oid, record->target_value)) < 0)
+			goto out;
+		ref = git_reference__alloc(record->ref_name, &oid, &peeled);
+	}
+	GIT_ERROR_CHECK_ALLOC(ref);
+
+	*out = ref;
+out:
+	return error;
+}
+
+static int reftable_log_fill(struct log_record *out,
+			     const git_signature *who,
+			     const git_oid *old_id,
+			     const git_oid *new_id,
+			     const char *reference,
+			     const char *message)
+{
+	memset(out, 0, sizeof(*out));
+	if (who) {
+		out->name = who->name;
+		out->email = who->email;
+		out->time = who->when.time;
+		out->tz_offset = who->when.offset;
+	}
+	if (old_id)
+		out->old_hash = (unsigned char *) old_id->id;
+	if (new_id)
+		out->new_hash = (unsigned char *) new_id->id;
+	out->ref_name = (char *) reference;
+	out->message = (char *) message;
+	return 0;
+}
+
+static int reftable_writer(struct writer *wr, void *_arg)
+{
+	reftable_writer_arg *arg = (reftable_writer_arg *) _arg;
+	writer_set_limits(wr, arg->index, arg->index + arg->nrefs + arg->nlogs);
```

Comment:
this should just be 

 writer_set_limits(wr, arg->index, arg->index)

the name update_index is misleading. It's better to think of it as "logical timestamp", and all the entries (both log & ref) should be written at the same update_index timestamp. Later, after compaction, this will makes it clear that these entries were written together.